sql/src/main/java/io/crate/execution/engine/indexing/ShardingUpsertExecutor.java

```diff
@@ -333,4 +336,22 @@ private void maybeSetInterrupt(@Nullable Exception failure) {
         }
     }
 
+    private static class HeapThresholdChecker implements Predicate<ShardedRequests<?, ?>> {
+
+        private final Runtime rt;
+        private final AtomicLong threshold;
+
+        HeapThresholdChecker(int freeHeapPercent) {
+            if (freeHeapPercent <= 0 || freeHeapPercent >= 100) {
+                throw new IllegalArgumentException(String.format("value must be between 0..100: %d", freeHeapPercent));
+            }
+            rt = Runtime.getRuntime();
+            threshold = new AtomicLong(rt.maxMemory() * freeHeapPercent / 100);
```

Comment:
The `AtomicLong` doesn't seem to be necessary, the threshold value is never mutated after the initial assignment.

Reply:
Correct, this is reminiscent of my first approach (somehow I wanted to make the default 30% 
 configurable and reloadable). Should I do that?